store/storage/rocksdb/db.go

```diff
@@ -0,0 +1,271 @@
+package rocksdb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"cosmossdk.io/store/v2"
+	"github.com/linxGnu/grocksdb"
+)
+
+const (
+	TimestampSize = 8
+
+	StorePrefixTpl   = "s/k:%s/"
+	latestVersionKey = "s/latest"
+)
+
+var (
+	_ store.VersionedDatabase = (*Database)(nil)
+
+	defaultWriteOpts = grocksdb.NewDefaultWriteOptions()
+	defaultReadOpts  = grocksdb.NewDefaultReadOptions()
+)
+
+type Database struct {
+	storage  *grocksdb.DB
+	cfHandle *grocksdb.ColumnFamilyHandle
+}
+
+func New(dataDir string) (*Database, error) {
+	storage, cfHandle, err := OpenRocksDB(dataDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open RocksDB: %w", err)
+	}
+
+	return &Database{
+		storage:  storage,
+		cfHandle: cfHandle,
+	}, nil
+}
+
+func NewWithDB(storage *grocksdb.DB, cfHandle *grocksdb.ColumnFamilyHandle) *Database {
+	return &Database{
+		storage:  storage,
+		cfHandle: cfHandle,
+	}
+}
+
+func (db *Database) Close() error {
+	db.storage.Close()
+	return nil
+}
+
+func (db *Database) getSlice(storeKey string, version uint64, key []byte) (*grocksdb.Slice, error) {
+	return db.storage.GetCF(
+		newTSReadOptions(version),
+		db.cfHandle,
+		prependStoreKey(storeKey, key),
+	)
+}
+
+func (db *Database) SetLatestVersion(version uint64) error {
+	var ts [TimestampSize]byte
+	binary.LittleEndian.PutUint64(ts[:], version)
+
+	return db.storage.Put(defaultWriteOpts, []byte(latestVersionKey), ts[:])
+}
+
+func (db *Database) GetLatestVersion() (uint64, error) {
+	bz, err := db.storage.GetBytes(defaultReadOpts, []byte(latestVersionKey))
+	if err != nil {
+		return 0, err
+	}
+
+	if len(bz) == 0 {
+		return 0, nil
+	}
+
+	return binary.LittleEndian.Uint64(bz), nil
```

Comment:
A few things:
1. When `len(bz) == 0` we aren't indicating to the caller that we didn't find the value, we are returning `(0, nil)` which is a paradoxical condition given that every single version is saved with a timestamp and value
2. This code is an instance where we should use the readOnly slice because all that we use the bytes for is to get the latest version as a uint64 and we aren't mutating the slice. We can invoke .GetCF as is and then invoke .Exists and then the readOnly slice to pass into binary.LittleEndian.Uint64. .GetBytes retrieves the slice then creates a copy of it assuming it might be mutated.

Reply:
- I've returned an error for the empty bytes case
- latest version isn't stored under a column family